### PR TITLE
[SPARK-49461][SS] Persistent Checkpoint ID to commit logs and read it back

### PR DESCRIPTION
**Pull request by WweiL**
### What changes were proposed in this pull request?

In this change, we propose to add a new field to the commit log if `STREAMING_STATE_STORE_COMMIT_LOG_VERSION` is 2. The new field is a `Map[String, Map[String, Map[String, Seq[String]]]]` that maps `OperatorId -> PartitionId -> StoreName -> Seq[uniqueId]`. This is a necessary step to enable the v2 checkpoint.

### Why are the changes needed?

New feature

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Added UT

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by WweiL** [comment]
cc @siying PTAL!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CommitLog.scala [hunk 1]
```diff
@@ -72,12 +74,37 @@ class CommitLog(sparkSession: SparkSession, path: String)
 }
 
 object CommitLog {
-  private val VERSION = 1
+  private val VERSION = SQLConf.get.stateStoreCheckpointFormatVersion
```
Comment (siying) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CommitLog.scala:77]: Why do we need to change this version?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/resources/testCommitLogV1/testCommitLog [hunk 2]
```diff
@@ -0,0 +1,2 @@
+v1
+{"nextBatchWatermarkMs":1,"stateUniqueIds":{}}
```
Comment (siying) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/resources/testCommitLogV1/testCommitLog:2]: It's better to put it as a constant string in the source code and write it in a file when the tests start. I am not sure we can always assume that when the test runs, we can access to the source code files. It is less flexible.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CommitLog.scala (hunk 1 above)
Comment (WweiL) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CommitLog.scala:77]: Currently we do not support switching back from checkpoint v2 to v1, the `validateVersion` function above ensures that.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/resources/testCommitLogV1/testCommitLog (hunk 2 above)
Comment (WweiL) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/resources/testCommitLogV1/testCommitLog:2]: This is the golden file generation mechanism, a pretty common pattern used in Spark: 
https://github.com/apache/spark/blob/master/core/src/test/scala/org/apache/spark/SparkFunSuite.scala#L91

See example usage here: https://github.com/apache/spark/blob/master/core/src/test/scala/org/apache/spark/SparkThrowableSuite.scala#L95

Comment (brkyvz) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/resources/testCommitLogV1/testCommitLog:2]: I assume this shouldn't have the `"stateUniqueIds":{}` field and should continue work without it

**Review by brkyvz — APPROVED** [review]
One comment, otherwise LGTM. I assume you don't want the extra field to show up in the V1 state and still have the code work for backwards compatibility

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/resources/testCommitLogV1/testCommitLog (hunk 2 above)
Comment (WweiL) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/resources/testCommitLogV1/testCommitLog:2]: This is for same-version verification, e.g. a spark 4.0 query (with this change) but have checkpoint format v2 turned off.

For cross version validation (a spark 4.0 query run against commit log created in spark 3.5, see the test here)
```
// Old metadata structure with no state unique ids should not affect the deserialization
  test("Cross-version V1 SerDe") {
    val commitlogV1 = """v1
                        |{"nextBatchWatermarkMs":233}""".stripMargin
    val inputStream: ByteArrayInputStream =
      new ByteArrayInputStream(commitlogV1.getBytes("UTF-8"))
    val commitMetadata: CommitMetadata = new CommitLog(
      spark, testCommitLogV1FilePath.toString).deserialize(inputStream)
    assert(commitMetadata.nextBatchWatermarkMs === 233)
    assert(commitMetadata.stateUniqueIds === Map.empty)
  }
```

**Review by brkyvz — APPROVED** [review]
LGTM!

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/resources/testCommitLogV1/testCommitLog (hunk 2 above)
Comment (brkyvz) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/resources/testCommitLogV1/testCommitLog:2]: ah my bad - was worried when I didn't see a golden file for it. Any reason you haven't created a golden file? Then you may not need to expose the methods inside CommitLog?

Comment (WweiL) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/resources/testCommitLogV1/testCommitLog:2]: nah just this way is easier...

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/CommitLogSuite.scala [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+      "resources",
+      "testCommitLogV1"
+    )
+  }
+
+  private def testSerde(commitMetadata: CommitMetadata, path: Path): Unit = {
+    if (regenerateGoldenFiles) {
+      val commitLog = new CommitLog(spark, path.toString)
+      val outputStream = new FileOutputStream(path.resolve("testCommitLog").toFile)
+      commitLog.serialize(commitMetadata, outputStream)
```
Comment (WweiL) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/CommitLogSuite.scala:67]: nah I still need to expose the serialize method here

**Comment by HeartSaVioR** [comment]
Same here, I'm merging the PR on behalf of @brkyvz as he asked personally. Just to leave DISCLAIMER.

**Comment by HeartSaVioR** [comment]
Thanks! Merging to master.

**Pull request closed without merging**